ompi/mca/coll/acoll/coll_acoll_allgather.c

```diff
@@ -0,0 +1,625 @@
+        }
+    } else if (!(rank & (sg_size - 1))) {
+        /* Use recursive doubling for power of 2 cases */
+        err = rd_allgather_sub(rbuf, rdtype, comm, bcount, brank, rank, brank, num_sgs, 1, sg_start,
+                               sg_size, rext);
+        if (MPI_SUCCESS != err) {
+            return err;
+        }
+    }
+    /* Now all base ranks have the full data */
+    /* Do broadcast within subgroups from the base ranks for the extra data */
+    if (sg_id == 0) {
+        num_data_blks = 1;
+        data_blk_size[0] = bcount * (num_sgs - 2) + last_subgrp_rcnt;
+        blk_ofst[0] = bcount;
+    } else if (sg_id == num_sgs - 1) {
+        if (last_subgrp_size < 2) {
+            return err;
+        }
+        num_data_blks = 1;
+        data_blk_size[0] = bcount * (num_sgs - 1);
+        blk_ofst[0] = 0;
+    } else {
+        num_data_blks = 2;
+        data_blk_size[0] = bcount * sg_id;
+        data_blk_size[1] = bcount * (num_sgs - sg_id - 2) + last_subgrp_rcnt;
+        blk_ofst[0] = 0;
+        blk_ofst[1] = bcount * (sg_id + 1);
+    }
+    reqs = ompi_coll_base_comm_get_reqs(module->base_data, size);
+    if (NULL == reqs) {
+        return OMPI_ERR_OUT_OF_RESOURCE;
+    }
+    nreqs = 0;
+    preq = reqs;
+    /* Loop over data blocks */
+    for (i = 0; i < num_data_blks; i++) {
+        char *buff = (char *) rbuf + (ptrdiff_t) blk_ofst[i] * rext;
+        int sg_dim = opal_hibit(subgrp_size - 1, comm->c_cube_dim);
+        if ((1 << sg_dim) < subgrp_size) {
+            sg_dim++;
+        }
+        /* The size parameters to sg_bcast_intra ensures that the no. of send
+           requests do not exceed the max allocated. */
+        err = sg_bcast_intra(buff, data_blk_size[i], rdtype, rank, sg_dim, size, sg_size, sg_start,
+                             sg_start, comm, module, preq, &nreqs);
+        if (MPI_SUCCESS != err) {
+            ompi_coll_base_free_reqs(reqs, nreqs);
+            return err;
+        }
+    }
+    /* Start and wait on all requests. */
+    if (nreqs > 0) {
+        err = ompi_request_wait_all(nreqs, reqs, MPI_STATUSES_IGNORE);
+        if (MPI_SUCCESS != err) {
+            ompi_coll_base_free_reqs(reqs, nreqs);
+        }
+    }
+
+    /* All done */
+    return err;
+}
+
+/*
+ * mca_coll_acoll_allgather
+ *
+ * Function:    Allgather operation using subgroup based algorithm
+ * Accepts:     Same arguments as MPI_Allgather()
+ * Returns:     MPI_SUCCESS or error code
+ *
+ * Description: Allgather is performed across and within subgroups.
+ *              Subgroups can be 1 or more based on size and count.
+ *
+ * Memory:      No additional memory requirements beyond user-supplied buffers.
+ *
+ */
+int mca_coll_acoll_allgather(const void *sbuf, size_t scount, struct ompi_datatype_t *sdtype,
+                             void *rbuf, size_t rcount, struct ompi_datatype_t *rdtype,
+                             struct ompi_communicator_t *comm, mca_coll_base_module_t *module)
```

Comment:
Overall I'm confused about the use of sbuf vs rbuf and MPI_IN_PLACE handling. Please see my other comments.